File: py/nexarda_sdk.py
# Nexarda SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import NexardaUtility
from core.spec import NexardaSpec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import NexardaBaseFeature
from features import _make_feature


class NexardaSDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = NexardaUtility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features from config.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            feature_items = vs.items(feature_opts)
            if feature_items is not None:
                for item in feature_items:
                    fname = item[0]
                    fopts = helpers.to_map(item[1])
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return NexardaUtility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = NexardaSpec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    @property
    def console(self):
        """Idiomatic facade: client.console.list() / client.console.load({"id": ...})."""
        from entity.console_entity import ConsoleEntity
        cached = getattr(self, "_console", None)
        if cached is None:
            cached = ConsoleEntity(self, None)
            self._console = cached
        return cached

    def Console(self, data=None):
        # Deprecated: use client.console instead.
        from entity.console_entity import ConsoleEntity
        return ConsoleEntity(self, data)


    @property
    def franchis(self):
        """Idiomatic facade: client.franchis.list() / client.franchis.load({"id": ...})."""
        from entity.franchis_entity import FranchisEntity
        cached = getattr(self, "_franchis", None)
        if cached is None:
            cached = FranchisEntity(self, None)
            self._franchis = cached
        return cached

    def Franchis(self, data=None):
        # Deprecated: use client.franchis instead.
        from entity.franchis_entity import FranchisEntity
        return FranchisEntity(self, data)


    @property
    def game(self):
        """Idiomatic facade: client.game.list() / client.game.load({"id": ...})."""
        from entity.game_entity import GameEntity
        cached = getattr(self, "_game", None)
        if cached is None:
            cached = GameEntity(self, None)
            self._game = cached
        return cached

    def Game(self, data=None):
        # Deprecated: use client.game instead.
        from entity.game_entity import GameEntity
        return GameEntity(self, data)


    @property
    def platform(self):
        """Idiomatic facade: client.platform.list() / client.platform.load({"id": ...})."""
        from entity.platform_entity import PlatformEntity
        cached = getattr(self, "_platform", None)
        if cached is None:
            cached = PlatformEntity(self, None)
            self._platform = cached
        return cached

    def Platform(self, data=None):
        # Deprecated: use client.platform instead.
        from entity.platform_entity import PlatformEntity
        return PlatformEntity(self, data)


    @property
    def price(self):
        """Idiomatic facade: client.price.list() / client.price.load({"id": ...})."""
        from entity.price_entity import PriceEntity
        cached = getattr(self, "_price", None)
        if cached is None:
            cached = PriceEntity(self, None)
            self._price = cached
        return cached

    def Price(self, data=None):
        # Deprecated: use client.price instead.
        from entity.price_entity import PriceEntity
        return PriceEntity(self, data)


    @property
    def retailer(self):
        """Idiomatic facade: client.retailer.list() / client.retailer.load({"id": ...})."""
        from entity.retailer_entity import RetailerEntity
        cached = getattr(self, "_retailer", None)
        if cached is None:
            cached = RetailerEntity(self, None)
            self._retailer = cached
        return cached

    def Retailer(self, data=None):
        # Deprecated: use client.retailer instead.
        from entity.retailer_entity import RetailerEntity
        return RetailerEntity(self, data)


    @property
    def search(self):
        """Idiomatic facade: client.search.list() / client.search.load({"id": ...})."""
        from entity.search_entity import SearchEntity
        cached = getattr(self, "_search", None)
        if cached is None:
            cached = SearchEntity(self, None)
            self._search = cached
        return cached

    def Search(self, data=None):
        # Deprecated: use client.search instead.
        from entity.search_entity import SearchEntity
        return SearchEntity(self, data)


    @property
    def studio(self):
        """Idiomatic facade: client.studio.list() / client.studio.load({"id": ...})."""
        from entity.studio_entity import StudioEntity
        cached = getattr(self, "_studio", None)
        if cached is None:
            cached = StudioEntity(self, None)
            self._studio = cached
        return cached

    def Studio(self, data=None):
        # Deprecated: use client.studio instead.
        from entity.studio_entity import StudioEntity
        return StudioEntity(self, data)


    @property
    def user(self):
        """Idiomatic facade: client.user.list() / client.user.load({"id": ...})."""
        from entity.user_entity import UserEntity
        cached = getattr(self, "_user", None)
        if cached is None:
            cached = UserEntity(self, None)
            self._user = cached
        return cached

    def User(self, data=None):
        # Deprecated: use client.user instead.
        from entity.user_entity import UserEntity
        return UserEntity(self, data)


    @property
    def widget(self):
        """Idiomatic facade: client.widget.list() / client.widget.load({"id": ...})."""
        from entity.widget_entity import WidgetEntity
        cached = getattr(self, "_widget", None)
        if cached is None:
            cached = WidgetEntity(self, None)
            self._widget = cached
        return cached

    def Widget(self, data=None):
        # Deprecated: use client.widget instead.
        from entity.widget_entity import WidgetEntity
        return WidgetEntity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None):
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk

File: lua/nexarda_sdk.lua
-- Nexarda SDK

local vs = require("utility.struct.struct")
local Utility = require("core.utility_type")
local Spec = require("core.spec")
local helpers = require("core.helpers")

-- Load utility registration (populates Utility._registrar)
require("utility.register")

-- Load features
local BaseFeature = require("feature.base_feature")
local features_factory = require("features")


local NexardaSDK = {}
NexardaSDK.__index = NexardaSDK


local function _make_feature(name)
  local factory = features_factory[name]
  if factory ~= nil then
    return factory()
  end
  return features_factory.base()
end

NexardaSDK._make_feature = _make_feature


function NexardaSDK.new(options)
  local self = setmetatable({}, NexardaSDK)
  self.mode = "live"
  self.features = {}
  self.options = nil

  local utility = Utility.new()
  self._utility = utility

  local config = require("config")()

  self._rootctx = utility.make_context({
    client = self,
    utility = utility,
    config = config,
    options = options or {},
    shared = {},
  }, nil)

  self.options = utility.make_options(self._rootctx)

  if vs.getpath(self.options, "feature.test.active") == true then
    self.mode = "test"
  end

  self._rootctx.options = self.options

  -- Add features from config.
  local feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
  if feature_opts ~= nil then
    local feature_items = vs.items(feature_opts)
    if feature_items ~= nil then
      for _, item in ipairs(feature_items) do
        local fname = item[1]
        local fopts = helpers.to_map(item[2])
        if fopts ~= nil and fopts["active"] == true then
          utility.feature_add(self._rootctx, _make_feature(fname))
        end
      end
    end
  end

  -- Add extension features.
  local extend = vs.getprop(self.options, "extend")
  if type(extend) == "table" then
    for _, f in ipairs(extend) do
      if type(f) == "table" and type(f.get_name) == "function" then
        utility.feature_add(self._rootctx, f)
      end
    end
  end

  -- Initialize features.
  for _, f in ipairs(self.features) do
    utility.feature_init(self._rootctx, f)
  end

  utility.feature_hook(self._rootctx, "PostConstruct")

  -- #BuildFeatures

  return self
end


function NexardaSDK:options_map()
  local out = vs.clone(self.options)
  if type(out) == "table" then
    return out
  end
  return {}
end


function NexardaSDK:get_utility()
  return Utility.copy(self._utility)
end


function NexardaSDK:get_root_ctx()
  return self._rootctx
end


function NexardaSDK:prepare(fetchargs)
  local utility = self._utility

  fetchargs = fetchargs or {}

  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "prepare",
    ctrl = ctrl,
  }, self._rootctx)

  local options = self.options

  local path = vs.getprop(fetchargs, "path") or ""
  if type(path) ~= "string" then path = "" end

  local method = vs.getprop(fetchargs, "method") or "GET"
  if type(method) ~= "string" then method = "GET" end

  local params = helpers.to_map(vs.getprop(fetchargs, "params")) or {}
  local query = helpers.to_map(vs.getprop(fetchargs, "query")) or {}

  local headers = utility.prepare_headers(ctx)

  local base = vs.getprop(options, "base") or ""
  if type(base) ~= "string" then base = "" end
  local prefix = vs.getprop(options, "prefix") or ""
  if type(prefix) ~= "string" then prefix = "" end
  local suffix = vs.getprop(options, "suffix") or ""
  if type(suffix) ~= "string" then suffix = "" end

  ctx.spec = Spec.new({
    base = base,
    prefix = prefix,
    suffix = suffix,
    path = path,
    method = method,
    params = params,
    query = query,
    headers = headers,
    body = vs.getprop(fetchargs, "body"),
    step = "start",
  })

  -- Merge user-provided headers.
  local uh = vs.getprop(fetchargs, "headers")
  if type(uh) == "table" then
    for k, v in pairs(uh) do
      ctx.spec.headers[k] = v
    end
  end

  local _, err = utility.prepare_auth(ctx)
  if err ~= nil then
    return nil, err
  end

  return utility.make_fetch_def(ctx)
end


function NexardaSDK:direct(fetchargs)
  local utility = self._utility

  local fetchdef, err = self:prepare(fetchargs)
  if err ~= nil then
    return { ok = false, err = err }, nil
  end

  fetchargs = fetchargs or {}
  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "direct",
    ctrl = ctrl,
  }, self._rootctx)

  local url = fetchdef["url"] or ""
  local fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

  if fetch_err ~= nil then
    return { ok = false, err = fetch_err }, nil
  end

  if fetched == nil then
    return {
      ok = false,
      err = ctx:make_error("direct_no_response", "response: undefined"),
    }, nil
  end

  if type(fetched) == "table" then
    local status = helpers.to_int(vs.getprop(fetched, "status"))
    local headers = vs.getprop(fetched, "headers") or {}

    -- No-body responses (204, 304) and explicit zero content-length
    -- must skip JSON parsing — calling json() on an empty body errors.
    local content_length = nil
    if type(headers) == "table" then
      content_length = headers["content-length"]
    end
    local no_body = status == 204 or status == 304 or tostring(content_length) == "0"

    local json_data = nil
    if not no_body then
      local jf = vs.getprop(fetched, "json")
      if type(jf) == "function" then
        local ok, result = pcall(jf)
        if ok then
          json_data = result
        end
        -- Non-JSON body: json_data stays nil, status/headers preserved.
      end
    end

    return {
      ok = status >= 200 and status < 300,
      status = status,
      headers = headers,
      data = json_data,
    }, nil
  end

  return {
    ok = false,
    err = ctx:make_error("direct_invalid", "invalid response type"),
  }, nil
end



-- Idiomatic facade: client:console():list() / client:console():load({ id = ... })
function NexardaSDK:console(data)
  local EntityMod = require("entity.console_entity")
  if data == nil then
    if self._console == nil then
      self._console = EntityMod.new(self, nil)
    end
    return self._console
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:console() instead.
function NexardaSDK:Console(data)
  local EntityMod = require("entity.console_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:franchis():list() / client:franchis():load({ id = ... })
function NexardaSDK:franchis(data)
  local EntityMod = require("entity.franchis_entity")
  if data == nil then
    if self._franchis == nil then
      self._franchis = EntityMod.new(self, nil)
    end
    return self._franchis
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:franchis() instead.
function NexardaSDK:Franchis(data)
  local EntityMod = require("entity.franchis_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:game():list() / client:game():load({ id = ... })
function NexardaSDK:game(data)
  local EntityMod = require("entity.game_entity")
  if data == nil then
    if self._game == nil then
      self._game = EntityMod.new(self, nil)
    end
    return self._game
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:game() instead.
function NexardaSDK:Game(data)
  local EntityMod = require("entity.game_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:platform():list() / client:platform():load({ id = ... })
function NexardaSDK:platform(data)
  local EntityMod = require("entity.platform_entity")
  if data == nil then
    if self._platform == nil then
      self._platform = EntityMod.new(self, nil)
    end
    return self._platform
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:platform() instead.
function NexardaSDK:Platform(data)
  local EntityMod = require("entity.platform_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:price():list() / client:price():load({ id = ... })
function NexardaSDK:price(data)
  local EntityMod = require("entity.price_entity")
  if data == nil then
    if self._price == nil then
      self._price = EntityMod.new(self, nil)
    end
    return self._price
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:price() instead.
function NexardaSDK:Price(data)
  local EntityMod = require("entity.price_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:retailer():list() / client:retailer():load({ id = ... })
function NexardaSDK:retailer(data)
  local EntityMod = require("entity.retailer_entity")
  if data == nil then
    if self._retailer == nil then
      self._retailer = EntityMod.new(self, nil)
    end
    return self._retailer
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:retailer() instead.
function NexardaSDK:Retailer(data)
  local EntityMod = require("entity.retailer_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:search():list() / client:search():load({ id = ... })
function NexardaSDK:search(data)
  local EntityMod = require("entity.search_entity")
  if data == nil then
    if self._search == nil then
      self._search = EntityMod.new(self, nil)
    end
    return self._search
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:search() instead.
function NexardaSDK:Search(data)
  local EntityMod = require("entity.search_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:studio():list() / client:studio():load({ id = ... })
function NexardaSDK:studio(data)
  local EntityMod = require("entity.studio_entity")
  if data == nil then
    if self._studio == nil then
      self._studio = EntityMod.new(self, nil)
    end
    return self._studio
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:studio() instead.
function NexardaSDK:Studio(data)
  local EntityMod = require("entity.studio_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:user():list() / client:user():load({ id = ... })
function NexardaSDK:user(data)
  local EntityMod = require("entity.user_entity")
  if data == nil then
    if self._user == nil then
      self._user = EntityMod.new(self, nil)
    end
    return self._user
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:user() instead.
function NexardaSDK:User(data)
  local EntityMod = require("entity.user_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:widget():list() / client:widget():load({ id = ... })
function NexardaSDK:widget(data)
  local EntityMod = require("entity.widget_entity")
  if data == nil then
    if self._widget == nil then
      self._widget = EntityMod.new(self, nil)
    end
    return self._widget
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:widget() instead.
function NexardaSDK:Widget(data)
  local EntityMod = require("entity.widget_entity")
  return EntityMod.new(self, data)
end




function NexardaSDK.test(testopts, sdkopts)
  sdkopts = sdkopts or {}
  sdkopts = vs.clone(sdkopts)
  if type(sdkopts) ~= "table" then
    sdkopts = {}
  end

  testopts = testopts or {}
  testopts = vs.clone(testopts)
  if type(testopts) ~= "table" then
    testopts = {}
  end
  testopts["active"] = true

  vs.setpath(sdkopts, "feature.test", testopts)

  local sdk = NexardaSDK.new(sdkopts)
  sdk.mode = "test"

  return sdk
end


return NexardaSDK
